### PR TITLE
Talon v0.4 support + some related improvements

### PR DESCRIPTION
**Pull request by knausj85**
- Convert many lists to .talon-list to jumpstart the process / discussion. We may need some sort of migration plan.
- Support for common commands when composing talon-list, .talon, and talon .py

Note: does not port to resource.watch (see https://github.com/talonhub/community/pull/1199)

Questions:
I believe we'll want a migration script for CSVs that are moved to .talon-list: reads the CSV, spits out a talon-list. 

- Should CSVs migrated to talon-lists migrated continue to live in Settings?
- do we want to auto regenerate a talon list if it's deleted as was done for the CSVs? 
- is there a 'good' way to migrate targeted lists that exist only in python without breaking every customization?

### CODE REVIEW

**Comment by knausj85** [comment]
note on talon-list conversion:
https://github.com/talonhub/community/blob/main/core/create_spoken_forms.py imports some things from keys, numbers, abbreviate and file extensions. As a result, I don't think these are necessarily good candidates for .talon-list conversion at the moment.

File: core/keys/function_key.talon-list [hunk 1]
```diff
@@ -0,0 +1,26 @@
+list: user.function_key
+-
+f one: f1
```
Comment (knausj85) [inline core/keys/function_key.talon-list:3]: It's a little unfortunate to have to change numbers in multiple places if people customize them. Thoughts?

File: core/keys/punctuation.talon-list [hunk 2]
```diff
@@ -0,0 +1,29 @@
+list: user.punctuation
```
Comment (knausj85) [inline core/keys/punctuation.talon-list:1]: This probably needs to be kept in python or moved to a csv for compatibility with create_spoken_forms

File: lang/talon/talon-python-activator.talon [hunk 3]
```diff
@@ -0,0 +1,7 @@
+app: vscode
+win.title: /— user/
+win.title: /— knausj_talon/
+win.title: / - user - Visual Studio Code/
+tag: user.python
+-
+tag(): user.talon_python
```
Comment (knausj85) [inline lang/talon/talon-python-activator.talon:7]: This is a breaking change to restrict talon-related python commands to relevant projects only. This is provided as an example for discussion.

File: lang/talon/talon.talon [hunk 4]
```diff
@@ -10,44 +10,6 @@ tag(): user.code_functions_common
...
-key <user.modifiers> over: "{modifiers}"
-
-# all actions (requires uncommenting user.talon_populate_lists tag above)
-funk {user.talon_actions}:
-    user.code_insert_function(talon_actions, edit.selected_text())
-funk cell <number>: user.code_select_function(number - 1, "")
-funk wrap <user.code_common_function>:
-    user.code_insert_function(code_common_function, edit.selected_text())
-funk wrap <number>: user.code_select_function(number - 1, edit.selected_text())
+action {user.talon_actions}:
```
Comment (knausj85) [inline lang/talon/talon.talon:14]: Note: if possible, I think the existing code_common_functions in talon.py should probably be migrated to user.talon_actions for those that can't take the perf hit of  

tag(): user.talon_populate_lists

and to better support both talon 
-specific actions and standard python functions.

File: lang/talon/talon-common.talon [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+key <user.keys> over: "{keys}"
+key <user.modifiers> over: "{modifiers}"
+
+# all actions (requires uncommenting user.talon_populate_lists tag above)
+
+funk cell <number>: user.talon_code_insert_function(number - 1, "")
+funk wrap <user.code_common_function>:
+    user.code_insert_function(code_common_function, edit.selected_text(), False)
+funk wrap <number>:
+    user.talon_code_insert_function(number - 1, edit.selected_text(), False)
```
Comment (knausj85) [inline lang/talon/talon-common.talon:26]: Note to self: need to take a fresh look at these. Why are these here?

File: core/modes/language_modes.py [hunk 6]
```diff
@@ -40,6 +40,7 @@
     # 'snippets': 'snippets',
     "sql": "sql",
     "talon": "talon",
+    "talonlist": "talon-list",
```
Comment (knausj85) [inline core/modes/language_modes.py:43]: Add a space for talon list spoken form

File: lang/talon/talon.py [hunk 7]
```diff
@@ -73,6 +73,17 @@ def on_ready():
 app.register("ready", on_ready)
 
 
+@mod.action_class
+class Actions:
+    def talon_code_insert_function(text: str, selection: str, prepend_actions: bool):
+        """inserts talent specific action with either user prepended or not"""
```
Comment (knausj85) [inline lang/talon/talon.py:79]: Correct doc string

File: core/keys/keypad_key.talon-list [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+key pad five: keypad_5
+key pad six: keypad_6
+key pad seven: keypad_7
+key pad eight: keypad_8
+key pad nine: keypad_9
+key pad point: keypad_decimal
+key pad plus: keypad_plus
+key pad minus: keypad_minus
+key pad times: keypad_multiply
+key pad stroke: keypad_divide
```
Comment (nriley) [inline core/keys/keypad_key.talon-list:17]: Perhaps use "slash" here to be consistent with keys.py?

File: core/keys/keypad_key.talon-list [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+key pad one: keypad_1
+key pad two: keypad_2
+key pad three: keypad_3
+key pad four: keypad_4
+key pad five: keypad_5
+key pad six: keypad_6
+key pad seven: keypad_7
+key pad eight: keypad_8
+key pad nine: keypad_9
+key pad point: keypad_decimal
```
Comment (nriley) [inline core/keys/keypad_key.talon-list:13]: Consider adding "dot" for consistency with keys.py.

File: core/keys/keypad_key.talon-list [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+key pad plus: keypad_plus
+key pad minus: keypad_minus
+key pad times: keypad_multiply
```
Comment (nriley) [inline core/keys/keypad_key.talon-list:16]: Seems odd that you have "times" but not "divided by". Would recommend either defaulting to the mathematical operations or the symbols inserted, but not a mixture of both.

File: core/modes/dictation/punctuation.talon-list [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+forward slash: /
+question mark: ?
+exclamation mark: !
+exclamation point: !
+asterisk: *
+number sign: #
+percent sign: %
+at sign: @
+and sign: &
+ampersand: &
```
Comment (nriley) [inline core/modes/dictation/punctuation.talon-list:23]: & is in here twice.

File: core/modes/dictation/punctuation.talon-list [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+# Currencies
+dollar sign: $
```
Comment (nriley) [inline core/modes/dictation/punctuation.talon-list:25]: $ is in here twice.

File: core/text/formatters.talon-list [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+dunder: dunder
+hammer: hammer
+kebab: kebab
+packed: packed
+padded: padded
+slasher: slasher
+smash: smash
+snake: snake
+string: string
+title: title
```
Comment (nriley) [inline core/text/formatters.talon-list:17]: This will need updating after #1216.

File: core/websites_and_search_engines/websites.talon-list [hunk 14]
```diff
@@ -0,0 +1,21 @@
+list: user.website
+-
+talon home page: http://talonvoice.com
+talon slack: https://app.slack.com/client/T7FPSMV8F/C7EK0P
+talon issues: https://github.com/talonvoice/talon/issues
+talon community: https://github.com/knausj85/knausj_talon
+talon community issues: https://github.com/knausj85/knausj_talon/issues
```
Comment (nriley) [inline core/websites_and_search_engines/websites.talon-list:7]: These will need updating with the talonhub/community URLs.

**Review by nriley — COMMENTED** [review]
This is by no means a complete review. If this doesn't create a complete disaster, it might be helpful to split this out into multiple PRs as it does so many different things.

File: apps/talon/talon_debug/talon_debug.talon [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+list {user.talon_lists}: "{talon_lists}"
+
+# requires user.talon_populate_lists tag. do not use with dragon
+capture {user.talon_captures}: "{talon_captures}"
+set {user.talon_settings}: user.paste("{talon_settings}")
+host:
+    hostname = user.talon_get_hostname()
+    user.paste("{hostname}")
+application {user.talon_apps}: user.paste("{talon_apps}")
+mode {user.talon_modes}: user.paste("{talon_modes}")
```
[the remark this answers is not in the document]
Comment (knausj85) [reply apps/talon/talon_debug/talon_debug.talon:20]: A lot of these are direct copies of existing commands that have been around for ages, as I couldn't come up with a good way to unify them all in single .talon file. they presently reflect whatever was used at the time. e.g., 
https://github.com/talonhub/community/blob/468fb16392e6a9907cb98e2526c1e5cbf3b5fc8d/lang/talon/talon.talon#L44

**Comment by knausj85** [comment]
> This is by no means a complete review. If this doesn't create a complete disaster, it might be helpful to split this out into multiple PRs as it does so many different things.

If you feel this is unwieldy as-is, I can look at splitting the review up. The branch has been underway for a bit, unfortunately

File: apps/talon/talon_debug/talon_debug.py
[the remark this answers is not in the document]
Comment (knausj85) [reply apps/talon/talon_debug/talon_debug.py:1]: Yep. I personally don't run Linux at all, so someone else would need to take a crack if they have time :)

**Comment by nriley** [comment]
> > This is by no means a complete review. If this doesn't create a complete disaster, it might be helpful to split this out into multiple PRs as it does so many different things.
> 
> If you feel this is unwieldy as-is, I can look at splitting the review up. The branch has been underway for a bit, unfortunately

Thanks, I do — hopefully it won't be too much trouble!
